internal/shim-sgx/src/handler.rs

```diff
@@ -159,126 +98,103 @@ impl<'a> Handler<'a> {
-        Ok([FAKE_UID.into(), 0.into()])
-    }
+    unsafe fn proxy(&mut self, req: Request) -> sallyport::Result {
+        self.block.msg.req = req;
 
-    /// Do a getegid() syscall
-    pub fn getegid(&mut self) -> sallyport::Result {
-        self.trace("getegid", 0);
-        Ok([FAKE_GID.into(), 0.into()])
-    }
+        // prevent earlier writes from being moved beyond this point
+        core::sync::atomic::compiler_fence(core::sync::atomic::Ordering::Release);
 
-    /// Do a read() syscall
-    pub fn read(&mut self) -> sallyport::Result {
-        self.trace("read", 3);
+        let _ret = syscall(self.aex, self.ctx);
 
-        let c = self.block.cursor();
-        let trusted_len: usize = self.aex.gpr.rdx.into();
-        let trusted: *mut u8 = self.aex.gpr.rsi.into();
-        let (_, untrusted) = c.alloc::<u8>(trusted_len).or(Err(libc::EMSGSIZE))?;
+        // prevent later reads from being moved before this point
+        core::sync::atomic::compiler_fence(core::sync::atomic::Ordering::Acquire);
 
-        let req = request!(libc::SYS_read => self.aex.gpr.rdi, untrusted, untrusted.len());
-        let ret = unsafe { self.proxy(req)? };
+        self.block.msg.rep.into()
+    }
 
-        if trusted_len < ret[0].into() {
-            self.attacked();
-        }
+    /// When we are under attack, we trip this circuit breaker and
+    /// exit the enclave. Any attempt to re-enter the enclave after
+    /// tripping the circuit breaker causes the enclave to immediately
+    /// EEXIT.
+    fn attacked(&mut self) -> ! {
+        self.exit(1)
+    }
 
-        let c = self.block.cursor();
-        unsafe { c.copy_into_raw_parts(trusted_len, trusted, ret[0].into()) }
-            .or(Err(libc::EMSGSIZE))?;
+    fn trace(&mut self, name: &str, argc: usize) {
+        if !TRACE {
+            return;
+        }
 
-        Ok(ret)
-    }
+        let argv = [
+            self.aex.gpr.rdi,
+            self.aex.gpr.rsi,
+            self.aex.gpr.rdx,
+            self.aex.gpr.r10,
+            self.aex.gpr.r8,
+            self.aex.gpr.r9,
+        ];
 
-    /// Do a write() syscall
-    pub fn write(&mut self) -> sallyport::Result {
-        self.trace("write", 3);
+        debug!(self, "{}(", name);
+        for (i, arg) in argv[..argc].iter().copied().enumerate() {
+            let prefix = if i > 0 { ", " } else { "" };
+            debug!(self, "{}0x{:x}", prefix, u64::from(arg));
+        }
 
-        let c = self.block.cursor();
-        let trusted: *mut u8 = self.aex.gpr.rsi.into();
-        let trusted_len: usize = self.aex.gpr.rdx.into();
-        let (_, untrusted) =
-            unsafe { c.copy_from_raw_parts(trusted, trusted_len) }.or(Err(libc::EMSGSIZE))?;
+        debugln!(self, ")");
+    }
 
-        let req = request!(libc::SYS_write => self.aex.gpr.rdi, untrusted, trusted_len);
-        let res = unsafe { self.proxy(req)? };
+    /// Proxy an exit() syscall
+    fn exit(&mut self, status: libc::c_int) -> ! {
```

Comment:
Since we don't access `self.aex.gpr.rdi` anymore directly, one can call the method with any parameter he/she wants (which is basically your old `code: T`=.